### ISSUE TITLE
feature: add progress?

### ISSUE THREAD

**Issue by tlambert03** [funkelab/ilpy#49]
look into  passing progress info on from the backend solver.  such as an iteration callback or something

**Comment by tlambert03** [comment]
SCIP offers events that we could hook into (https://scipopt.org/doc/html/group__PublicEventMethods.php) and presumably gurobi has similar.

Alternatively, we could just parse stdout (after setting `solve.set_verbose(True)`)

**Comment by tlambert03** [comment]
just for records: here's an example of verbose output from Scip backend:

```
➜ python x.py
feasible solution found by trivial heuristic after 0.0 seconds, objective value 0.000000e+00
presolving:
(round 1, fast)       3 del vars, 12 del conss, 0 add conss, 0 chg bounds, 0 chg sides, 0 chg coeffs, 37 upgd conss, 0 impls, 0 clqs
(round 2, fast)       5 del vars, 17 del conss, 0 add conss, 0 chg bounds, 4 chg sides, 4 chg coeffs, 37 upgd conss, 0 impls, 5 clqs
   (0.0s) running MILP presolver
   (0.0s) MILP presolver (2 rounds): 0 aggregations, 0 fixings, 0 bound changes
(round 3, exhaustive) 5 del vars, 19 del conss, 0 add conss, 0 chg bounds, 4 chg sides, 4 chg coeffs, 37 upgd conss, 0 impls, 5 clqs
(round 4, exhaustive) 5 del vars, 19 del conss, 0 add conss, 0 chg bounds, 4 chg sides, 4 chg coeffs, 59 upgd conss, 0 impls, 5 clqs
(round 5, medium)     6 del vars, 31 del conss, 24 add conss, 0 chg bounds, 5 chg sides, 5 chg coeffs, 59 upgd conss, 0 impls, 28 clqs
(round 6, fast)       6 del vars, 32 del conss, 24 add conss, 0 chg bounds, 5 chg sides, 5 chg coeffs, 59 upgd conss, 0 impls, 28 clqs
(round 7, medium)     7 del vars, 33 del conss, 24 add conss, 0 chg bounds, 5 chg sides, 5 chg coeffs, 59 upgd conss, 0 impls, 27 clqs
(round 8, exhaustive) 7 del vars, 33 del conss, 24 add conss, 0 chg bounds, 5 chg sides, 5 chg coeffs, 66 upgd conss, 0 impls, 27 clqs
(round 9, medium)     7 del vars, 37 del conss, 32 add conss, 0 chg bounds, 13 chg sides, 25 chg coeffs, 66 upgd conss, 0 impls, 35 clqs
(round 10, exhaustive) 7 del vars, 62 del conss, 39 add conss, 0 chg bounds, 13 chg sides, 25 chg coeffs, 66 upgd conss, 0 impls, 35 clqs
(round 11, exhaustive) 7 del vars, 66 del conss, 39 add conss, 0 chg bounds, 13 chg sides, 25 chg coeffs, 66 upgd conss, 0 impls, 35 clqs
   (0.0s) probing cycle finished: starting next cycle
   (0.0s) symmetry computation started: requiring (bin +, int -, cont +), (fixed: bin -, int +, cont -)
   (0.0s) no symmetry present (symcode time: 0.00)
presolving (12 rounds: 12 fast, 9 medium, 6 exhaustive):
 7 deleted vars, 66 deleted constraints, 39 added constraints, 0 tightened bounds, 0 added holes, 13 changed sides, 25 changed coefficients
 0 implications, 53 cliques
presolved problem has 23 variables (23 bin, 0 int, 0 impl, 0 cont) and 22 constraints
     13 constraints of type <setppc>
      7 constraints of type <and>
      1 constraints of type <linear>
      1 constraints of type <logicor>
transformed objective value is always integral (scale: 1)
Presolving Time: 0.00
transformed 1/1 original solutions to the transformed problem space

 time | node  | left  |LP iter|LP it/n|mem/heur|mdpt |vars |cons |rows |cuts |sepa|confs|strbr|  dualbound   | primalbound  |  gap   | compl.
  0.0s|     1 |     0 |    10 |     - |  1343k |   0 |  23 |  28 |  29 |   0 |  0 |   5 |   0 |-2.805000e+02 | 0.000000e+00 |    Inf | unknown
r 0.0s|     1 |     0 |    10 |     - |shifting|   0 |  23 |  23 |  29 |   0 |  0 |   5 |   0 |-2.805000e+02 |-1.120000e+02 | 150.45%| unknown
  0.0s|     1 |     0 |    16 |     - |  1351k |   0 |  23 |  23 |  31 |   2 |  1 |   5 |   0 |-2.305000e+02 |-1.120000e+02 | 105.80%| unknown
r 0.0s|     1 |     0 |    16 |     - |rounding|   0 |  23 |  23 |  31 |   2 |  1 |   5 |   0 |-2.305000e+02 |-2.060000e+02 |  11.89%| unknown
  0.0s|     1 |     0 |    16 |     - |  1354k |   0 |  23 |  23 |  31 |   2 |  1 |   5 |   0 |-2.305000e+02 |-2.060000e+02 |  11.89%| unknown
  0.0s|     1 |     0 |    17 |     - |  1354k |   0 |  23 |  24 |  24 |   2 |  1 |   6 |   0 |-2.296667e+02 |-2.060000e+02 |  11.49%| unknown
  0.0s|     1 |     0 |    18 |     - |  1356k |   0 |  23 |  22 |  25 |   3 |  2 |   6 |   0 |-2.060000e+02 |-2.060000e+02 |   0.00%| unknown
  0.0s|     1 |     0 |    18 |     - |  1356k |   0 |  23 |  22 |  25 |   3 |  2 |   6 |   0 |-2.060000e+02 |-2.060000e+02 |   0.00%| unknown

SCIP Status        : problem is solved [optimal solution found]
Solving Time (sec) : 0.02
Solving Nodes      : 1
Primal Bound       : -2.06000000000000e+02 (4 solutions)
Dual Bound         : -2.06000000000000e+02
Gap                : 0.00 %

```

and from gurobi:

```

CPU model: Apple M2 Max
Thread count: 12 physical cores, 12 logical processors, using up to 1 threads

Optimize a model with 0 rows, 30 columns and 0 nonzeros
Model fingerprint: 0xfb958892
Model has 49 quadratic constraints
Variable types: 0 continuous, 30 integer (30 binary)
Coefficient statistics:
  Matrix range     [0e+00, 0e+00]
  QLMatrix range   [1e+00, 2e+00]
  Objective range  [1e+00, 2e+02]
  Bounds range     [1e+00, 1e+00]
  RHS range        [0e+00, 0e+00]
  QRHS range       [1e+00, 2e+00]
Found heuristic solution: objective 0.0000000
Presolve added 22 rows and 0 columns
Presolve removed 0 rows and 7 columns
Presolve time: 0.00s
Presolved: 22 rows, 23 columns, 63 nonzeros
Variable types: 0 continuous, 23 integer (23 binary)

Root relaxation: objective -2.550000e+02, 8 iterations, 0.00 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0 -255.00000    0    6    0.00000 -255.00000      -     -    0s
H    0     0                    -206.0000000 -255.00000  23.8%     -    0s

Explored 1 nodes (8 simplex iterations) in 0.00 seconds (0.00 work units)
Thread count was 1 (of 12 available processors)

Solution count 2: -206 0
No other solutions better than -206

Optimal solution found (tolerance 1.00e-04)
Best objective -2.060000000000e+02, best bound -2.060000000000e+02, gap 0.0000%
```

(prolly need a more complex problem to see more from gurobi)

**Comment by tlambert03** [comment]
also a note to self.  here's a hacky script that will run the solver in another thread, capture output, and print the gap as it comes.  Can be used as a starting point:

```python
import os
import queue
import re
import sys
import threading

import ilpy
import motile
from motile.constraints import MaxChildren, MaxParents
from motile.costs import Appear, EdgeDistance, EdgeSelection, NodeSelection, Split
from motile.data import arlo_graph

graph = arlo_graph()

solver = motile.Solver(graph)
solver.add_costs(NodeSelection(weight=-1.0, attribute="score", constant=-100.0))
solver.add_costs(
    EdgeSelection(weight=0.5, attribute="prediction_distance", constant=-1.0)
)
solver.add_costs(EdgeDistance(position_attributes=("x",), weight=0.5))
solver.add_costs(Appear(constant=200.0, attribute="score", weight=-1.0))
solver.add_costs(Split(constant=100.0, attribute="score", weight=1.0))

solver.add_constraints(MaxParents(1))
solver.add_constraints(MaxChildren(2))


def capture_output(q, func, *args, **kwargs):
    # Save the original stdout file descriptor
    original_stdout_fd = sys.stdout.fileno()

    # Create a duplicate of the original stdout file descriptor
    saved_stdout_fd = os.dup(original_stdout_fd)

    # Create a temporary file and replace the original stdout file descriptor with it
    with open("temp.txt", "w") as f:
        os.dup2(f.fileno(), original_stdout_fd)

    # Call the function
    result = func(*args, **kwargs)

    # Restore the original stdout file descriptor using the duplicate
    os.dup2(saved_stdout_fd, original_stdout_fd)

    # Read the output from the temporary file and put it in the queue
    with open("temp.txt") as f:
        for line in f:
            q.put(line)

    q.put(None)  # Signal that we're done
    return result




def log_output(q: queue.Queue) -> None:
    # Get lines from the queue and log them
    for line in iter(q.get, None):
        if any(x in line for x in {"probing", "best"}):
            continue
        # dumb way to get the gap, it's the only thing expressed as a percentage
        if match := re.search(r"(\d+(\.?\d+))%", line):
            print(match.group(1))


q = queue.Queue()
t1 = threading.Thread(
    target=capture_output,
    args=(q, solver.solve),
    kwargs={"pref": ilpy.Preference.Scip, "verbose": True},
)
t2 = threading.Thread(target=log_output, args=(q,))

t1.start()
t2.start()

t1.join()
t2.join()
```

**Comment by cmalinmayor** [comment]
Resolved by #56
